src/mscorlib/shared/System/IO/FileStream.cs

```diff
@@ -292,6 +292,36 @@ public override Task FlushAsync(CancellationToken cancellationToken)
             return FlushAsyncInternal(cancellationToken);
         }
 
+        public override int Read(byte[] array, int offset, int count)
+        {
+            ValidateReadWriteArgs(array, offset, count);
+            return _useAsyncIO ?
+                ReadAsyncInternal(array, offset, count, CancellationToken.None).GetAwaiter().GetResult() :
+                ReadSpan(new Span<byte>(array, offset, count));
+        }
+
+        public override int Read(Span<byte> destination)
+        {
+            if (GetType() == typeof(FileStream) && !_useAsyncIO)
```

Comment:
Faster for what?

The fast-path here requires checking both, as we'll write the span directly if this is a concrete FileStream and we're not configured in async mode.  If either of those conditions is false, we have to take a slower path, and both slower paths involve so much additional overhead that I don't think it matters which we check first.  If this is a type derived from FileStream, then we end up getting a temporary array from the array pool, making another virtual call to do the read into that array, copying the contents of the array back into the span, and then returning the array to the pool.  And if the stream is configured for async mode but you're doing a sync read on it, we do everything I just said, plus we do the read using overlapped I/O, allocating the associated state, and block an entire thread waiting for it to complete.  A small amount of that we may be able to get rid of once we have `Buffer<T>`, but most of it will remain.  Compared to that, I don't think it matters whether we check the bool or the type first, but I can swap them if you care.